src/docker/composeappengine.cc

```diff
@@ -193,4 +320,77 @@ void ComposeAppEngine::extractAppArchive(const App& app, const std::string& arch
   }
 }
 
+ComposeAppEngine::AppState::AppState(const App& app, const boost::filesystem::path& root, bool store_version) try
+    : version_file_ {
+  (root / MetaDir / VersionFile).string()
+}
+, state_file_{(root / MetaDir / StateFile).string()} {
+  version_ = version_file_.readStr();
+
+  if (version_ != app.uri) {
+    if (store_version) {
+      version_ = app.uri;
+      version_file_.write(version_);
+      setState(AppState::State::kUnknown);
+    }
+    state_ = AppState::State::kUnknown;
+    return;
+  }
+
+  state_ = static_cast<State>(state_file_.read());
+}
+catch (const std::exception& exc) {
+  LOG_ERROR << "Failed to read version or state file: " << exc.what();
+}
+
+void ComposeAppEngine::AppState::setState(const State& state) {
+  state_ = state;
+  state_file_.write(static_cast<int>(state));
+}
+
+ComposeAppEngine::AppState::File::File(const std::string& filepath) : filepath_{filepath} {
+  fd_ = open(filepath.c_str(), O_CREAT | O_RDWR | O_SYNC, S_IRUSR | S_IWUSR);
+  if (fd_ == -1) {
+    throw std::runtime_error("Failed to open/create a version file, " + filepath + ", error: " + strerror(errno));
+  }
+}
+
+ComposeAppEngine::AppState::File::~File() { close(fd_); }
+
+void ComposeAppEngine::AppState::File::write(int val) {
+  lseek(fd_, 0, SEEK_SET);
+  ssize_t wr = ::write(fd_, &val, sizeof(val));
+  if (wr == -1) {
+    throw std::runtime_error("Failed to write to file, " + filepath_ + ", error: " + strerror(errno));
+  }
+  fsync(fd_);
+}
+
+void ComposeAppEngine::AppState::File::write(const std::string& val) {
+  lseek(fd_, 0, SEEK_SET);
+  ssize_t wr = ::write(fd_, val.data(), val.size());
+  if (wr == -1) {
+    throw std::runtime_error("Failed to write to file, " + filepath_ + ", error: " + strerror(errno));
+  }
```

Comment:
if the length of `val` shrinks, I think you'll need a call to something like `ftruncate`. but it seems like a safer/simpler approach would be writing to a tempfile and then doing a rename of the file. this would prevent a user making a concurrent call to read the file from getting inconsistent data.

Reply:
I was planning to do the renaming stuff, this is still WIP.

Reply:
done